src/NServiceBus.Core/Recoverability/FailureInfoStorage.cs

```diff
@@ -0,0 +1,122 @@
+namespace NServiceBus
+{
+    using System.Collections.Generic;
+    using System.Runtime.ExceptionServices;
+    using JetBrains.Annotations;
+
+    // The data structure has fixed maximum size. When the data structure reaches its maximum size,
+    // the least recently used (LRU) message processing failure is removed from the storage.
+    class FailureInfoStorage
```

Comment:
If we don't want to share this between `FLR`, `SLR` and `Errors` can we make this generic? `SLR` and `Errors` would use `FailureInfoStorage.Of(ExceptionDispatchInfo)` and `FLR` `FailureInfoStorage.Of(Tuple.Of(ExceptionDispatchInfo, int)).